lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -1,5 +1,105 @@
+        }
+        {
+          type: "variable.parameter"
+          value: "equation"
+        }
+        {
+          type: "rparen"
+          value: "}"
+        }
+      ]
+      END_EQUATION_TOKEN_SEQUENCE: [
+        {
+          type: "storage.type"
+          value: "\\end"
+        }
+        {
+          type: "lparen"
+          value: "{"
+        }
+        {
+          type: "variable.parameter"
+          value: "equation"
+        }
+        {
+          type: "rparen"
+          value: "}"
+        }
+      ]
+
+      equalTokens: (token1, token2) ->
+        if token1? and token2?
+          return token1.type == token2.type and token1.value == token2.value
+        else return if token1? or token2? then false else true
+
+      getEquationStart: (tokenIterator) ->
+        # following cycle pushes tokenIterator to the end of
+        # beginning sequence, if it is inside one
+        for token in erh.BEGIN_EQUATION_TOKEN_SEQUENCE
+          if erh.equalTokens(token, tokenIterator.getCurrentToken())
+            tokenIterator.stepForward()
+        curSequenceIndex = erh.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+        curEquationStart = null
+        while curSequenceIndex >= 0
+          if erh.equalTokens(
+              erh.BEGIN_EQUATION_TOKEN_SEQUENCE[curSequenceIndex],
+              tokenIterator.stepBackward())
+            if curSequenceIndex == erh.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+              curTokenPosition = tokenIterator.getCurrentTokenPosition()
+              curEquationStart = {
+                row: curTokenPosition.row
+                column: curTokenPosition.column + tokenIterator.getCurrentToken().value.length
+              }
+            curSequenceIndex -= 1
+          else
+            curSequenceIndex = erh.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+            curEquationStart = null
+        return curEquationStart
+
+      getEquationEnd: (tokenIterator) ->
+        # following cycle pushes tokenIterator to the start of
+        # ending sequence, if it is inside one
+        for token in erh.END_EQUATION_TOKEN_SEQUENCE.slice(0).reverse()
+          if erh.equalTokens(token, tokenIterator.getCurrentToken())
+            tokenIterator.stepBackward()
+        curSequenceIndex = 0
+        curEquationStart = null
+        while curSequenceIndex < erh.END_EQUATION_TOKEN_SEQUENCE.length
+          if erh.equalTokens(
+            erh.END_EQUATION_TOKEN_SEQUENCE[curSequenceIndex],
+            tokenIterator.stepForward())
+            if curSequenceIndex == 0
+              curEquationStart = tokenIterator.getCurrentTokenPosition()
+            curSequenceIndex += 1
+          else
+            curSequenceIndex = 0
+            curEquationStart = null
+        return curEquationStart
+
+      getEquationRange: (row, column) ->
+        lineLength = editor.getSession().getLine(row).length
+        start = erh.getEquationStart(new TokenIterator(editor.getSession(), row, lineLength))
```

Comment:
This is to handle cases like this:

    Hi there! \begin{equation} ...

However, I've just realized we can avoid this witchcraft, if we just search for end first, and then search start from there.